plugin-cgroupv2/src/k8s/token.rs

```diff
@@ -0,0 +1,140 @@
+                None => Err(anyhow!("could no read the token's value")),
+            };
+        }
+
+        self.refresh().await
+    }
+
+    /// Retrieves the k8s API token using either a kubectl command
+    /// or by reading  the service account token's file.
+    async fn refresh(self: Arc<Self>) -> anyhow::Result<String> {
+        let token = match self.token_retrieval {
+            TokenRetrieval::Kubectl => {
+                let output = Command::new("kubectl")
+                    .args(["create", "token", "alumet-reader", "-n", "alumet"])
+                    .output()?;
+
+                let token = String::from_utf8_lossy(&output.stdout);
+                token.trim().to_string()
+            }
+            TokenRetrieval::File => {
+                let mut file = File::open("/var/run/secrets/kubernetes.io/serviceaccount/token")?;
+                let mut token = String::new();
+                file.read_to_string(&mut token)?;
+                token
+            }
+        };
+
+        let mut components = token.split('.');
+        let _ = components.next().ok_or(anyhow!("Could not parse the token"))?;
+        let payload: serde_json::Value = serde_json::from_slice(
+            &BASE64_STANDARD_NO_PAD.decode(components.next().ok_or(anyhow!("Could not parse the token"))?.trim())?,
+        )?;
+        let _ = components.next().ok_or(anyhow!("Could not parse the token"))?;
+
+        if components.next().is_some() {
+            return Err(anyhow!("Token has too many components"));
+        }
+
+        // Extract 'exp' from the payload
+        let exp = match payload.get("exp") {
+            Some(exp) => exp.as_u64().unwrap_or_default(),
+            None => return Err(anyhow!("Could not extract the 'exp' field from the token")),
+        };
+
+        {
+            let mut new_exp = self.expiration_time.write().await;
+            *new_exp = Some(exp);
+        }
+
+        {
+            let mut new_value = self.value.write().await;
+            *new_value = Some(token.clone());
+        }
+
+        Ok(token)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use time::Duration;
+    use tokio::time;
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test_is_valid() {
+        let token: Token = Token::new(TokenRetrieval::File);
+        assert!(!token.is_valid().await);
+
+        {
+            let mut new_value = token.value.write().await;
+            *new_value = Some("abcde".to_string());
+        }
+
+        {
+            let mut new_exp = token.expiration_time.write().await;
+            *new_exp = Some(
+                SystemTime::now()
+                    .checked_add(Duration::from_secs(2))
+                    .unwrap()
+                    .duration_since(UNIX_EPOCH)
+                    .unwrap()
+                    .as_secs(),
+            );
+        }
+
+        assert!(token.is_valid().await);
+
+        let _ = time::sleep(Duration::from_secs(2)).await;
```

Comment:
Could you use a faster sleep here? 2 seconds is quite long for one test.

Reply:
No more sleep! 🩲